### PR TITLE
add more methods and tests for reductions over empty arrays

### PR DESCRIPTION
**Pull request by simonbyrne**
From discussion in #28535.

### CODE REVIEW

File: base/reduce.jl [hunk 1]
```diff
@@ -318,6 +318,9 @@ reduce_empty(::typeof(*), ::Type{<:AbstractChar}) = ""
 reduce_empty(::typeof(&), ::Type{Bool}) = true
 reduce_empty(::typeof(|), ::Type{Bool}) = false
 
+reduce_empty(::typeof(max), ::Type{<:AbstractFloat}) where {T} = typemin(T)
+reduce_empty(::typeof(min), ::Type{<:AbstractFloat}) where {T} = typemax(T)
```
Comment (vtjnash) [inline base/reduce.jl:322]:
```suggestion
reduce_empty(::typeof(max), T::Type{<:AbstractFloat}) = typemin(T)
reduce_empty(::typeof(min), T::Type{<:AbstractFloat}) = typemax(T)
```

**Review by vtjnash — APPROVED** [review]
Can you update this so that the doctest passes?

**Comment by StefanKarpinski** [comment]
Bump. Needs doctest fixes. Otherwise good to go.

**Comment by timholy** [comment]
Now that we have `maximum(itr; init=typemin(eltype(itr)))`, is this really the way we want to go? What if someone wanted the error? The `init` approach lets the user/developer choose.

#41885 will print a nice warning suggesting to supply an `init` value.

**Comment by timholy** [comment]
But :+1: to more tests!

**Comment by tkf** [comment]
Is it OK to get `false` from `-1 <= maximum(sin.(xs)) <= 1`? (I prefer an error)

**Comment by oscardssmith** [comment]
Let's figure out if this is OK to merge and then merge it.

File: base/reduce.jl [hunk 2]
```diff
@@ -346,8 +349,10 @@ mapreduce_empty(::typeof(identity), op, T) = reduce_empty(op, T)
 mapreduce_empty(::typeof(abs), op, T)      = abs(reduce_empty(op, T))
 mapreduce_empty(::typeof(abs2), op, T)     = abs2(reduce_empty(op, T))
 
-mapreduce_empty(f::typeof(abs),  ::typeof(max), T) = abs(zero(T))
-mapreduce_empty(f::typeof(abs2), ::typeof(max), T) = abs2(zero(T))
+mapreduce_empty(::typeof(abs),  ::typeof(max), T) = abs(zero(T))
+mapreduce_empty(::typeof(abs),  ::typeof(min), T::Type{<:AbstractFloat}) = typemax(T)
```
Comment (tkf) [inline base/reduce.jl:353]: Why not support complex, integers, etc.?

```suggestion
mapreduce_empty(::typeof(abs),  ::typeof(min), T) = typemax(abs(zero(T)))
```

File: base/reduce.jl [hunk 3]
```diff
@@ -346,8 +349,10 @@ mapreduce_empty(::typeof(identity), op, T) = reduce_empty(op, T)
...
+mapreduce_empty(::typeof(abs2), ::typeof(max), T) = abs2(zero(T))
+mapreduce_empty(::typeof(abs2),  ::typeof(min), T::Type{<:AbstractFloat}) = typemax(T)
```
Comment (tkf) [inline base/reduce.jl:355]:
```suggestion
mapreduce_empty(::typeof(abs2),  ::typeof(min), T) = typemax(abs2(zero(T)))
```

File: base/reduce.jl [hunk 4]
```diff
@@ -318,6 +318,9 @@ reduce_empty(::typeof(*), ::Type{<:AbstractChar}) = ""
 reduce_empty(::typeof(&), ::Type{Bool}) = true
 reduce_empty(::typeof(|), ::Type{Bool}) = false
 
+reduce_empty(::typeof(max), T::Type{<:AbstractFloat}) = typemin(T)
+reduce_empty(::typeof(min), T::Type{<:AbstractFloat}) = typemax(T)
```
Comment (tkf) [inline base/reduce.jl:322]: Why not just let `typemin`/`typemax` throw an error `MethodError: no method matching typemax(::Type{T})`?

```suggestion
reduce_empty(::typeof(max), T) = typemin(T)
reduce_empty(::typeof(min), T) = typemax(T)
```

**Review by tkf — COMMENTED** [review]
FWIW, I'm opposed to this change. We don't know the domain on which the user data is defined and we don't know if `typemin` and `typemax` are meaningful values (see an example above). I think nudging users to specify the identity element via `init` will help them write more robust programs.

That said, I think the following tweaks might make sense if we go with this.

**Comment by JeffBezanson** [comment]
Triage thinks it's ok to add these, with tkf's suggestions. `maximum(sin.(xs))` where `xs` is empty is unfortunate, but all you're doing is passing an empty array to `maximum`, so it doesn't matter how you got it. `maximum(sin, xs)` is different, and should remain an error.

**Pull request merged**